sunpy/image/transform.py

```diff
@@ -167,7 +107,175 @@ def _get_transform_method(method, use_scipy):
+    return wrapper
+
+
+def _handle_nan_missing_externally(rotation_function):
+    """
+    Decorator for a rotation function to externally handle NaN as the value for `missing`
+    """
+    @wraps(rotation_function)
+    def wrapper(image, matrix, shift, order, missing, clip):
+        # Sanitize missing if it is NaN
+        missing_to_use = np.nanmin(image) if np.isnan(missing) else missing
+
+        rotated_image = rotation_function(image, matrix, shift, order, missing_to_use, clip)
+
+        if np.isnan(missing):
+            # Rotate a constant image to determine where to apply the NaNs for `missing`
+            constant = scipy.ndimage.affine_transform(np.ones_like(image).T.astype(int), matrix,
+                                                      offset=shift, order=0, mode='constant').T
+            rotated_image[constant < 1] = np.nan
+
+        return rotated_image
+
+    return wrapper
+
+
+# Define individual rotation implementations below and insert in the dictionary at the end.
+#
+# The arguments must be, in order:
+#   image : `numpy.ndarray`
+#       The image, which could be integers or floats, and can include NaNs
+#   matrix : `numpy.ndarray` that is 2x2
+#       The linear transformation matrix (e.g., rotation+scale+skew)
+#   shift : 2-element `numpy.ndarray`
+#       The translational shift to apply to the image in each axis
+#   order : `int`
+#       The numerical parameter that controls the degree of interpolation
+#   missing : `float`
+#       The value to use for outside the bounds of the original image, and may be NaN
+#   clip : `bool`
+#       Whether to clip the output image to the range of the input image
+#
+# There are three decorators that can be used:
+#
+# * `_handle_clipping_externally` is for rotation functions that do not implement clipping
+# * `_handle_nan_missing_externally` is for rotation functions that do not handle NaN for `missing`
+# * `_handle_image_nans_externally` is for rotation functions that do not handle NaNs in the image
+#
+# Do not modify `image` directly; copy first if necessary.
+
+
+@_handle_clipping_externally
+@_handle_image_nans_externally
+def _rotation_scipy(image, matrix, shift, order, missing, clip):
+    """
+    Rotation using SciPy
+    """
+    rotated_image = scipy.ndimage.affine_transform(image.T, matrix, offset=shift, order=order,
+                                                   mode='constant', cval=missing).T
+
+    return rotated_image
+
+
+@_handle_clipping_externally
+@_handle_nan_missing_externally
+@_handle_image_nans_externally
+def _rotation_skimage(image, matrix, shift, order, missing, clip):
+    """
+    Rotation using scikit-image
+
+    We handle clipping externally because the clipping behavior of
+    :func:`skimage.transform.warp` is different between orders in {0, 1, 3} and
+    orders in {2, 4, 5}.
+
+    We handle NaN for `missing` externally because :func:`skimage.transform.warp`
+    does not handle it correctly for orders in {2, 4, 5}.
```

Comment:
These docstrings are useful but are in these hidden functions. If we had a `@sunpy_affine_transform` decorator, the other thing it could do is take the docstring out of here and put it in the `affine_transform` docstring to have details about why you might want to pick one over the other.